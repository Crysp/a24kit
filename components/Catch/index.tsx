import React from 'react';
import { H } from '../Typography';
import {
    Wrapper,
    Header,
    FlashIcon,
    Text,
} from './styled';

export interface Props {
    children: React.ReactNode
}

interface State {
    hasError: boolean
    code: null | string
}

export default class Catch extends React.Component<Props, State> {
    state = {
        hasError: false,
        code: null,
    };

    componentDidCatch() {
        this.setState({
            hasError: true,
        });
    }

    render() {
        const {
            hasError,
            code,
        } = this.state;
        const { children } = this.props;
        if (hasError) {
            return (
                <Wrapper>
                    <Header>
                        <FlashIcon />
                        <H size={4}>Что-то пошло не так...</H>
                    </Header>
                    <Text>
                        Мы уже знаем об этой проблеме. Очень скоро мы ее исправим.
                    </Text>
                    {code !== null && (
                        <Text>
                            <span>Код ошибки</span>
                            <span>: </span>
                            <b>{code}</b>
                        </Text>
                    )}
                </Wrapper>
            );
        }
        return children;
    }
}
