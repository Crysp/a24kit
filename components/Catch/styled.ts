import styled from 'styled-components';
import { Flash } from '../Icon';
import { Color } from '../config';

export const Wrapper = styled.div`
    padding: 20px;
    background: ${Color.White};
    color: ${Color.Coal};
    border-radius: 2px;
    border: 2px solid ${Color.Pale};
    font: 400 14px 'Open Sans', sans-serif;
`;

export const Header = styled.div`
    display: flex;
    align-items: center;
`;

export const FlashIcon = styled(Flash)`
    width: 20px;
    height: 20px;
    margin-right: 8px;
    fill: ${Color.Red};
`;

export const Text = styled.div`
    margin-top: 12px;
`;
