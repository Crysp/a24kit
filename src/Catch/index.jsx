import React from 'react';
import PropTypes from 'prop-types';
import { H } from '../Typography';
import {
    Wrapper,
    Header,
    FlashIcon,
    Text,
} from './styled';


/**
 * @develop
 */
export default class Catch extends React.Component {
    static propTypes = {
        children: PropTypes.node.isRequired,
    };

    state = {
        hasError: false,
        code: null,
    };

    componentDidCatch(error, errorInfo) {
        this.setState({
            hasError: true,
        });
    }

    render() {
        const {
            hasError,
            code,
        } = this.state;
        const { children } = this.props;
        if (hasError) {
            return (
                <Wrapper>
                    <Header>
                        <FlashIcon />
                        <H size={4}>Что-то пошло не так...</H>
                    </Header>
                    <Text>
                        Мы уже знаем об этой проблеме. Очень скоро мы ее исправим.
                    </Text>
                    {code !== null && (
                        <Text>
                            <span>Код ошибки</span>
                            <span>: </span>
                            <b>{code}</b>
                        </Text>
                    )}
                </Wrapper>
            );
        }
        return children;
    }
}
