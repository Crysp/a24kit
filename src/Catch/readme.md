```js
class BuggyCounter extends React.Component {
  constructor(props) {
    super(props);
    this.state = { counter: 0 };
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    this.setState(({counter}) => ({
      counter: counter + 1
    }));
  }
  render() {
    if (this.state.counter === 5) {
      // Simulate a JS error
      throw new Error('I crashed!');
    }
    return <Button onClick={this.onClick}>Count {this.state.counter}</Button>;
  }
}

<Catch>
    <BuggyCounter/>
</Catch>
```