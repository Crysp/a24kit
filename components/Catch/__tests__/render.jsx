import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import Catch from '../index';

// eslint-disable-next-line react/prop-types
function Buggy({ hasError = false }) {
    if (hasError) {
        throw new Error('Bug');
    }
    return <div>Полезный контент</div>;
}


describe('Компонент <Catch/>', () => {
    const consoleError = console.error;
    beforeEach(() => {
        console.error = () => {};
    });
    it('Дефолтный', () => {
        const tree = renderer.create(<Catch><Buggy /></Catch>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('С ошибкой', () => {
        const tree = renderer.create(<Catch><Buggy hasError /></Catch>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('С кодом ошибки', () => {
        const catchComponent = renderer.create(<Catch><Buggy hasError /></Catch>);
        catchComponent.getInstance().setState({ code: 'awesomeErrorCode' });
        expect(catchComponent.toJSON()).toMatchSnapshot();
    });
    afterEach(() => {
        console.error = consoleError;
    });
});
