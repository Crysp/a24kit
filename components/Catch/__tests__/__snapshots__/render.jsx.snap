// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент <Catch/> Дефолтный 1`] = `
<div>
  Полезный контент
</div>
`;

exports[`Компонент <Catch/> С кодом ошибки 1`] = `
.c4 {
  margin: 0;
  font-family: "Open Sans",sans-serif;
  color: inherit;
  font-size: 16px;
  line-height: 24px;
  font-weight: 600;
}

.c4:empty {
  display: inline-block;
  width: 40%;
  height: 16px;
  border-radius: 2px;
  background-image: linear-gradient( 90deg, #e1e1e1, #e1e1e1 30%, #ececec, #e1e1e1 70%, #e1e1e1 );
  background-size: 500px;
  -webkit-animation: MORgY .8s linear infinite;
  animation: MORgY .8s linear infinite;
}

.c3 {
  display: inline-block;
  fill: currentColor;
  stroke: currentColor;
  stroke-width: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 100%;
  height: 100%;
}

.c0 {
  padding: 20px;
  background: #ffffff;
  color: #333333;
  border-radius: 2px;
  border: 2px solid #e1e1e1;
  font: 400 14px 'Open Sans',sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  width: 20px;
  height: 20px;
  margin-right: 8px;
  fill: #e95454;
}

.c5 {
  margin-top: 12px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <svg
      aria-hidden="true"
      className="c2 c3"
      focusable="false"
      size="auto"
      style={Object {}}
      viewBox="0 0 6 18"
    >
      <path
        d="M3.94 7l-1.49.49 1.43-5.23a1 1 0 0 0-.7-1.23 1 1 0 0 0-1.23.71L.04 8.75c-.1.36 0 .74.27.99s.65.34 1 .23l1.56-.51-.75 3.15a1 1 0 0 0 .97 1.23 1 1 0 0 0 .98-.77l1.15-4.89a.98.98 0 0 0-.29-.96A1 1 0 0 0 3.94 7zm-2.02 9.44c0 .26.11.53.29.72.19.18.44.29.71.29.26 0 .52-.11.71-.29.18-.19.29-.46.29-.72 0-.27-.11-.52-.29-.71a1.05 1.05 0 0 0-1.42 0c-.18.19-.29.45-.29.71z"
      />
    </svg>
    <div
      className="c4"
      size={4}
    >
      Что-то пошло не так...
    </div>
  </div>
  <div
    className="c5"
  >
    Мы уже знаем об этой проблеме. Очень скоро мы ее исправим.
  </div>
  <div
    className="c5"
  >
    <span>
      Код ошибки
    </span>
    <span>
      : 
    </span>
    <b>
      awesomeErrorCode
    </b>
  </div>
</div>
`;

exports[`Компонент <Catch/> С ошибкой 1`] = `
.c4 {
  margin: 0;
  font-family: "Open Sans",sans-serif;
  color: inherit;
  font-size: 16px;
  line-height: 24px;
  font-weight: 600;
}

.c4:empty {
  display: inline-block;
  width: 40%;
  height: 16px;
  border-radius: 2px;
  background-image: linear-gradient( 90deg, #e1e1e1, #e1e1e1 30%, #ececec, #e1e1e1 70%, #e1e1e1 );
  background-size: 500px;
  -webkit-animation: MORgY .8s linear infinite;
  animation: MORgY .8s linear infinite;
}

.c3 {
  display: inline-block;
  fill: currentColor;
  stroke: currentColor;
  stroke-width: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 100%;
  height: 100%;
}

.c0 {
  padding: 20px;
  background: #ffffff;
  color: #333333;
  border-radius: 2px;
  border: 2px solid #e1e1e1;
  font: 400 14px 'Open Sans',sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  width: 20px;
  height: 20px;
  margin-right: 8px;
  fill: #e95454;
}

.c5 {
  margin-top: 12px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <svg
      aria-hidden="true"
      className="c2 c3"
      focusable="false"
      size="auto"
      style={Object {}}
      viewBox="0 0 6 18"
    >
      <path
        d="M3.94 7l-1.49.49 1.43-5.23a1 1 0 0 0-.7-1.23 1 1 0 0 0-1.23.71L.04 8.75c-.1.36 0 .74.27.99s.65.34 1 .23l1.56-.51-.75 3.15a1 1 0 0 0 .97 1.23 1 1 0 0 0 .98-.77l1.15-4.89a.98.98 0 0 0-.29-.96A1 1 0 0 0 3.94 7zm-2.02 9.44c0 .26.11.53.29.72.19.18.44.29.71.29.26 0 .52-.11.71-.29.18-.19.29-.46.29-.72 0-.27-.11-.52-.29-.71a1.05 1.05 0 0 0-1.42 0c-.18.19-.29.45-.29.71z"
      />
    </svg>
    <div
      className="c4"
      size={4}
    >
      Что-то пошло не так...
    </div>
  </div>
  <div
    className="c5"
  >
    Мы уже знаем об этой проблеме. Очень скоро мы ее исправим.
  </div>
</div>
`;
